lib/ide/src/transport/web.rs

```diff
@@ -0,0 +1,250 @@
+    /// The connection is open and ready to communicate.
+    Open,
+    /// The connection is in the process of closing.
+    Closing,
+    /// The connection is closed or couldn't be opened.
+    Closed,
+    /// Any other, unknown condition.
+    Unknown(u16),
+}
+
+impl State {
+    /// Returns current state of the given WebSocket.
+    pub fn query_ws(ws:&web_sys::WebSocket) -> State {
+        State::from_code(ws.ready_state())
+    }
+
+    /// Translates code returned by `WebSocket.readyState` into our enum.
+    /// cf https://developer.mozilla.org/en-US/docs/Web/API/WebSocket/readyState
+    pub fn from_code(code:u16) -> State {
+        match code {
+            web_sys::WebSocket::CONNECTING => State::Connecting,
+            web_sys::WebSocket::OPEN       => State::Open,
+            web_sys::WebSocket::CLOSING    => State::Closing,
+            web_sys::WebSocket::CLOSED     => State::Closed,
+            num                            => State::Unknown(num), // impossible
+        }
+    }
+}
+
+
+
+// =================
+// === WebSocket ===
+// =================
+
+/// Wrapper over JS `WebSocket` object and callbacks to its signals.
+#[derive(Debug)]
+pub struct WebSocket {
+    /// Handle to the JS `WebSocket` object.
+    pub ws         : web_sys::WebSocket,
+    /// Handle to a closure connected to `WebSocket.onmessage`.
+    pub on_message : ClosureStorage<MessageEvent>,
+    /// Handle to a closure connected to `WebSocket.onclose`.
+    pub on_close   : ClosureStorage<CloseEvent>,
+    /// Handle to a closure connected to `WebSocket.onopen`.
+    pub on_open    : ClosureStorage<Event>,
+    /// Handle to a closure connected to `WebSocket.onerror`.
+    pub on_error   : ClosureStorage<Event>,
+}
+
+impl WebSocket {
+    /// Wraps given WebSocket object.
+    pub fn new(ws:web_sys::WebSocket) -> WebSocket {
+        WebSocket {
+            ws,
+            on_message : default(),
+            on_close   : default(),
+            on_open    : default(),
+            on_error   : default(),
+        }
+    }
+
+    /// Establish connection with endpoint defined by the given URL and wrap it.
+    /// Asynchronous, because it waits until connection is established.
+    pub async fn new_opened(url:impl Str) -> Result<WebSocket,ConnectingError> {
+        let     ws  = web_sys::WebSocket::new(url.as_ref());
+        let mut wst = WebSocket::new(ws.map_err(|e| {
+            ConnectingError::ConstructionError(js_to_string(e))
+        })?);
+
+        wst.wait_until_open().await?;
+        Ok(wst)
+    }
+
+    /// Awaits until `open` signal has been emitted. Clears any callbacks on
+    /// this `WebSocket`, if any has been set.
+    async fn wait_until_open(&mut self) -> Result<(),ConnectingError> {
+        // Connecting attempt shall either emit on_open or on_close.
+        // We shall wait for whatever comes first.
+        let (tx, mut rx) = mpsc::unbounded::<Result<(),()>>();
+        let tx_clone = tx.clone();
```

Comment:
if this is a cheap clone, please use `clone_ref`

Reply:
What is "cheap clone" ? 
Compiler says there's no such method. Am I missing some import?